File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-7.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        if nums[0] < nums[-1]:
            return nums[0]
        
        left, right, mid = 0, len(nums) - 1, len(nums) // 2

        while left <= right:
            if nums[mid - 1] > nums[mid]:
                return nums[mid]
            if nums[left] > nums[mid]:
                right = mid
            elif nums[right] < nums[mid]:
                left = mid
            # else:
            #     return nums[0]

            mid = (left + right + 1) // 2

File: Data Structures & Algorithms/invert-a-binary-tree/submission-7.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode invertTree(TreeNode root) {

        if (root == null) {
            return root;
        }
        
        Stack<TreeNode> s = new Stack();

        s.push(root);

        while (!s.isEmpty()) {
            TreeNode node = s.pop();

            TreeNode temp = node.left;
            node.left = node.right;
            node.right = temp;

            if (node.left != null) {
                s.push(node.left);
            }
            if (node.right != null) {
                s.push(node.right);
            }
        }

        return root;
    }
}
